testing/junit/src/main/java/com/linecorp/armeria/testing/junit/server/mockwebserver/MockWebServerExtension.java

```diff
@@ -0,0 +1,205 @@
+import com.linecorp.armeria.common.HttpResponseWriter;
+import com.linecorp.armeria.server.Server;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.testing.junit.server.ServerExtension;
+
+/**
+ * An {@link Extension} primarily for testing clients. {@link MockWebServerExtension} will start and stop a
+ * {@link Server} at the beginning and end of a test class. Call
+ * {@link MockWebServerExtension#enqueue(MockResponse)} as many times as you will make requests
+ * within the test. The enqueued responses will be returned in order for each of these requests. Later, call
+ * {@link MockWebServerExtension#takeRequest()} to retrieve requests in the order the server received them to
+ * validate the request parameters.
+ *
+ * <p>Note, tests in a class using {@link MockWebServerExtension} cannot be run concurrently, i.e., do not set
+ * {@link org.junit.jupiter.api.parallel.Execution} to
+ * {@link org.junit.jupiter.api.parallel.ExecutionMode#CONCURRENT}.
+ *
+ * <p>For example, <pre>{@code
+ *
+ * class MyTest {
+ *
+ * {@literal @}RegisterExtension
+ * static MockWebServerExtension server = new MockWebServerExtension();
+ *
+ * {@literal @}Test
+ * void checkSomething() {
+ *     HttpClient client = HttpClient.of(server.httpUri("/"));
+ *
+ *     server.enqueue(AggregatedHttpResponse.of(HttpStatus.OK));
+ *     server.enqueue(AggregatedHttpResponse.of(HttpStatus.FORBIDDEN));
+ *
+ *     assertThat(client.get("/").aggregate().join().status()).isEqualTo(HttpStatus.OK);
+ *     assertThat(client.get("/bad").aggregate().join().status()).isEqualTo(HttpStatus.FORBIDDEN);
+ *
+ *     assertThat(server.takeRequest().path()).isEqualTo("/");
+ *     assertThat(server.takeRequest().path()).isEqualTo("/bad");
+ * }
+ *
+ * }
+ * }</pre>
+ */
+public class MockWebServerExtension extends ServerExtension implements BeforeTestExecutionCallback {
+
+    private final Queue<MockResponse> mockResponses = new LinkedBlockingQueue<>();
+    private final Queue<RecordedRequest> recordedRequests = new LinkedBlockingQueue<>();
+
+    /**
+     * Enqueues the {@link MockResponse} to return to a client of this {@link MockWebServerExtension}. Multiple
+     * calls will return multiple responses in order.
+     */
+    public MockWebServerExtension enqueue(MockResponse response) {
+        mockResponses.add(response);
+        return this;
+    }
+
+    /**
+     * Enqueues the {@link AggregatedHttpResponse} to return to a client of this {@link MockWebServerExtension}.
+     * Multiple calls will return multiple responses in order.
+     */
+    public MockWebServerExtension enqueue(AggregatedHttpResponse response) {
+        mockResponses.add(MockResponse.of(response));
+        return this;
+    }
+
+    /**
+     * Returns the next {@link RecordedRequest} the server received. Call this method multiple times to retrieve
```

Comment:
I would make this blocking and update docs accordingly. it is much easier to write tests with blocking gets vs having strange awaitility wrapped race conditions

Reply:
Yup meant to block, fixed.